src/plugins/auth/ldap.js

```diff
@@ -139,11 +140,96 @@ function ldapAuth(manager, client, user, password, callback) {
 	return auth(user, password, callbackWrapper);
 }
 
+/**
+ * Use the LDAP filter from config to check that users still exist before loading them
+ * via the supplied callback function.
+ */
+
+function advancedLdapLoadUsers(users, callbackLoadUser) {
+	const config = Helper.config;
+
+	const ldapclient = ldap.createClient({
+		url: config.ldap.url,
+		tlsOptions: config.ldap.tlsOptions,
+	});
+
+	const base = config.ldap.searchDN.base;
+
+	ldapclient.on("error", function (err) {
+		log.error(`Unable to connect to LDAP server: ${err}`);
+	});
+
+	ldapclient.bind(config.ldap.searchDN.rootDN, config.ldap.searchDN.rootPassword, function (err) {
+		if (err) {
+			log.error("Invalid LDAP root credentials");
+		} else {
+			const userMap = {};
+
+			users.forEach((user) => (userMap[user] = false));
+
+			const searchOptions = {
+				scope: config.ldap.searchDN.scope,
+				filter: `${config.ldap.searchDN.filter}`,
+				attributes: [config.ldap.primaryKey],
+				paged: true,
+			};
+
+			ldapclient.search(base, searchOptions, function (err2, res) {
+				if (err2) {
+					log.error(`LDAP search error: ${err2}`);
+				} else {
+					res.on("searchEntry", function (entry) {
+						const user = entry.attributes[0]._vals[0].toString();
```

Comment:
This seems error prone, why doesn't `advancedLdapAuth` do this?

Reply:
Because ldapjs returns DNs (which advancedLdapAuth uses) as strings in objectName, but attributes (such as 'uid') as buffers:

2020-04-23 13:18:36 [DEBUG] entry:  SearchEntry {
  messageID: 2,
  protocolOp: 100,
  controls: [],
  log: { child: [Function (anonymous)] },
  objectName: 'cn=Genevieve,ou=accounts,dc=localhost',
  attributes: [ Attribute { type: 'uid', _vals: [Array] } ],

or :

2020-04-23 13:21:10 [DEBUG] entry.attributes:  [ Attribute { type: 'uid', _vals: [ <Buffer 78 61 76 69 65 72> ] } ]

So the best we could do here is to check that the attribute type is the same as the filter requested, but that seems redundant?

Reply:
I guess `attributes[0]` will match `attributes: [config.ldap.primaryKey]` ?

Reply:
yes. I can put a check in here if you think that it's worth it, but as I say it seems redundant to me,